### PR TITLE
Adds a new Provider for self-hosted atlassian

### PR DESCRIPTION
**Pull request by heiglandreas**
This commit adds a new provider for self-hosted attlasian-instances. It has been tested against a self-hosted confluence version 5.9.8.

Due to missing API-versioning there might be adaptions needed for general usage. Currently I only have this instance available for testing

There currently also are *no* tests!

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/SocialConnect/auth/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change:

Thanks :smiley_cat:

### CODE REVIEW

**Comment by heiglandreas** [comment]
Bugs fixed. Though no new tests… As I coudln't find some examples for other OAuth1-Providers I didn't add some…

File: src/OAuth1/Provider/Atlassian.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+        return $hydrator->hydrate(new User(), $result);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function oauthRequest($uri, $method = Client::GET, $parameters = [])
+    {
+        error_log($uri);
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: Debug things should be removed

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+use SocialConnect\OAuth1\Token;
+use SocialConnect\OAuth1\Util;
+
+class MethodRSASHA1 extends AbstractSignatureMethod
+{
+    private $private_key;
+
+    public function __construct($private_key)
+    {
+        if (! is_readable($private_key)) {
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: unexpected spaces

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+namespace SocialConnect\OAuth1\Signature;
+
+use SocialConnect\Provider\Consumer;
+use SocialConnect\OAuth1\Request;
+use SocialConnect\OAuth1\Token;
+use SocialConnect\OAuth1\Util;
+
+class MethodRSASHA1 extends AbstractSignatureMethod
+{
+    private $private_key;
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: not camelCase

File: src/OAuth1/Provider/Atlassian.php (hunk 1 above)
Comment (heiglandreas) [inline src/OAuth1/Provider/Atlassian.php]: ouch! Right! :(

File: src/OAuth1/Signature/MethodRSASHA1.php (hunk 2 above)
Comment (heiglandreas) [inline src/OAuth1/Signature/MethodRSASHA1.php]: where?

File: src/OAuth1/Signature/MethodRSASHA1.php (hunk 3 above)
Comment (heiglandreas) [inline src/OAuth1/Signature/MethodRSASHA1.php]: I think I remove that one and use the ```applicationSecret```-parameter for that… 'Cause that's what it is in the end…

File: src/OAuth1/Provider/Atlassian.php (hunk 1 above)
Comment (heiglandreas) [inline src/OAuth1/Provider/Atlassian.php]: That method shouldn't be there in the first place as it is obsoleted by #36

File: src/OAuth1/AbstractProvider.php [hunk 4]
```diff
@@ -45,6 +45,12 @@
     protected $scope = [];
 
     /**
+     * @var  \SocialConnect\OAuth1\Signature\AbstractSignatureMethod
+     */
+    protected $signature;
+
+
```
Comment (ovr) [inline src/OAuth1/AbstractProvider.php]: unneeded blank line

File: src/OAuth1/Provider/Atlassian.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
+<?php
+/**
+ * SocialConnect project
+ * @author: Patsura Dmitry https://github.com/ovr <talk@dmtry.me>
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: you @author

File: src/OAuth1/Provider/Atlassian.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+     * * private_key The path to the private key file used for signing.
+     *
+     * @param \SocialConnect\Common\Http\Client\ClientInterface $httpClient
+     * @param \SocialConnect\Provider\Session\SessionInterface  $session
+     * @param \SocialConnect\Provider\Consumer           $consumer
+     * @param array                                             $parameters
+     */
+    public function __construct(ClientInterface $httpClient, SessionInterface $session, Consumer $consumer, array $parameters)
+    {
+        if (! isset($parameters['baseUri'])) {
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: `if (! i` -> `if (!i` - please remove unneeded space

File: src/OAuth1/Provider/Atlassian.php [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+use SocialConnect\Provider\Consumer;
+use SocialConnect\Provider\Exception\InvalidResponse;
+use SocialConnect\OAuth1\AbstractProvider;
+use SocialConnect\Common\Entity\User;
+use SocialConnect\Common\Hydrator\ObjectMap;
+use SocialConnect\Provider\Session\SessionInterface;
+
+class Atlassian extends AbstractProvider
+{
+    private $baseUri;
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: blank line after `private $baseUri;`
and please doc block that it's a string
Thank

File: src/OAuth1/Provider/Atlassian.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        if (!$response->isSuccess()) {
+            throw new InvalidResponse(
+                'API response with error code',
+                $response
+            );
+        }
+
+        $headers = $response->getHeader('X-AUSERNAME');
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: $headers is unused, please remove it, thanks

File: src/OAuth1/Provider/Atlassian.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            'oauth_consumer_key' => $this->consumer->getKey(),
+            'oauth_token' => $this->consumerToken->getToken(),
+        ];
+
+        $response = $this->oauthRequest(
+            $this->getBaseUri() . '/rest/prototype/1/user/current',
+            Client::GET,
+            $parameters
+        );
+        while ($response->hasHeader('Location')) {
```
Comment (ovr) [inline src/OAuth1/Provider/Atlassian.php]: plank line before while
and btw, this while can be an infinity loop
if you are sure that attlasin can lag, maybe add retryCount, and retryMax? Thank

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * SocialConnect project
+ * @author: Patsura Dmitry https://github.com/ovr <talk@dmtry.me>
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: You @author

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+use SocialConnect\OAuth1\Token;
+use SocialConnect\OAuth1\Util;
+
+class MethodRSASHA1 extends AbstractSignatureMethod
+{
+    private $privateKey;
+
+    public function __construct($privateKey)
+    {
+        if (! is_readable($privateKey)) {
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: `if (! is_readable($privateKey))` - please remove uneeded space
Thank

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+namespace SocialConnect\OAuth1\Signature;
+
+use SocialConnect\Provider\Consumer;
+use SocialConnect\OAuth1\Request;
+use SocialConnect\OAuth1\Token;
+use SocialConnect\OAuth1\Util;
+
+class MethodRSASHA1 extends AbstractSignatureMethod
+{
+    private $privateKey;
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: Docblock that it's string?

File: src/OAuth1/Signature/MethodRSASHA1.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    public function __construct($privateKey)
+    {
+        if (! is_readable($privateKey)) {
+            throw new \InvalidArgumentException('The private key is not readable');
+        }
+
+        if (! function_exists('openssl_pkey_get_private')) {
```
Comment (ovr) [inline src/OAuth1/Signature/MethodRSASHA1.php]: `if (! function_exists('openssl_pkey_get_private')) {` - please remove unneeded space
Thank

File: tests/Test/OAuth1/Provider/AtlassianTest.php [hunk 14]
```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Copyright (c) Andreas Heigl<andreas@heigl.org>
```
Comment (ovr) [inline tests/Test/OAuth1/Provider/AtlassianTest.php]: It's not a SocialConnect LICENSE header

File: tests/Test/OAuth1/Signature/MethodRSASHA1Test.php [hunk 15]
```diff
@@ -0,0 +1,83 @@
+<?php
+/**
+ * Copyright (c) Andreas Heigl<andreas@heigl.org>
```
Comment (ovr) [inline tests/Test/OAuth1/Signature/MethodRSASHA1Test.php]: Again, not SociallConnect LICENSE header

**Comment by ovr** [comment]
@heiglandreas I added small comments, if you can fix this will be cool
and if it this will be merged
Thank

**Comment by heiglandreas** [comment]
All your comments should be addressed by now. Thanks for the feedback!

File: src/OAuth1/Provider/Atlassian.php [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    {
+        return 'atlassian';
+    }
+
+    /**
+     * Atlassian constructor.
+     *
+     * Required configuration parameters are:
+     *
+     * * base_uri The base URI of your self-hosted atlassian product
```
Comment (ADmad) [inline src/OAuth1/Provider/Atlassian.php]: In code the key is `baseUri` not `base_uri`.

File: src/OAuth1/Provider/Atlassian.php [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+     * * private_key The path to the private key file used for signing.
```
Comment (ADmad) [inline src/OAuth1/Provider/Atlassian.php]: Is this correct? A quick search doesn't show any use of `$parameters['private_key']` in the code.

Comment (heiglandreas) [inline src/OAuth1/Provider/Atlassian.php]: Thanks for spotting!!! I'll check that!!

**Comment by ovr** [comment]
BIG Thank @heiglandreas Awesome! Take it :beer: 
Thank @ADmad for review

**Pull request merged**